crates/bcr-ebill-transport/src/handler/company_chain_event_processor.rs

```diff
@@ -0,0 +1,792 @@
+    async fn add_company_blocks(
+        &self,
+        company_id: &NodeId,
+        chain: &mut CompanyBlockchain,
+        blocks: Vec<CompanyBlock>,
+    ) -> Result<()> {
+        let keys = self
+            .company_store
+            .get_key_pair(company_id)
+            .await
+            .map_err(|e| Error::Persistence(e.to_string()))?;
+        let mut company = self
+            .company_store
+            .get(company_id)
+            .await
+            .map_err(|e| Error::Persistence(e.to_string()))?;
+
+        let mut block_height = chain.get_latest_block().id;
+        for block in blocks {
+            if block.id <= block_height {
+                info!(
+                    "Skipping block with id {block_height} for {company_id} as we already have it"
+                );
+                continue;
+            }
+            let data = block
+                .get_block_data(&keys)
+                .map_err(|e| Error::Blockchain(e.to_string()))?;
+            if chain.try_add_block(block.clone()) {
+                block_height = block.id;
+                company.apply_block_data(&data);
+                self.save_block(company_id, &block).await?;
+                self.company_store
+                    .update(company_id, &company)
+                    .await
+                    .map_err(|e| Error::Persistence(e.to_string()))?;
+
+                if let CompanyBlockPayload::AddSignatory(payload) = data {
+                    self.nostr_contact_processor
+                        .ensure_nostr_contact(&payload.signatory)
+                        .await
+                };
+            }
+        }
+        debug!("Updated company {company_id} with data from new blocks");
+        Ok(())
+    }
+
+    fn get_valid_chain(
+        &self,
+        blocks: Vec<CompanyBlock>,
+        keys: &CompanyKeys,
+    ) -> Result<(NodeId, Company, CompanyBlockchain)> {
+        match CompanyBlockchain::new_from_blocks(blocks) {
+            Ok(chain) if chain.is_chain_valid() => {
+                let first_block = chain.get_first_block();
+                // create block is where we build up the company from
+                let payload = match first_block
+                    .get_block_data(keys)
+                    .map_err(|e| Error::Blockchain(e.to_string()))?
+                {
+                    CompanyBlockPayload::Create(payload) => payload,
+                    _ => {
+                        error!("First block of newly received company chain is not a Create block");
+                        return Err(Error::Blockchain(
+                            "First block of newly received company chain is not a Create block"
+                                .to_string(),
+                        ));
+                    }
+                };
+
+                // extracted node_id
+                let node_id = payload.id.clone();
+
+                // initialize company from payload
+                let mut company = Company {
```

Comment:
This could be a creator on the `Company` from `CompanyBlockPayload::Create` to be consistent with the Update/Signatory stuff in the `apply` function

Reply:
Made it a normal constructor from block. The apply method requires a self to be present.

Reply:
Yeah that's what I mean :+1: just consistent in terms of where the code lives